crates/steel-core/src/parser/expander.rs

```diff
@@ -625,251 +677,235 @@ impl MacroPattern {
-                            SyntaxObject {
-                                ty: TokenType::BooleanLiteral(s),
-                                ..
-                            },
-                    }) if s == b => continue,
-                    _ => return false,
-                },
-                MacroPattern::IntLiteral(i) => match val {
-                    ExprKind::Atom(Atom {
-                        syn:
-                            SyntaxObject {
-                                ty:
-                                    TokenType::Number(NumberLiteral::Real(RealLiteral::Int(
-                                        IntLiteral::Small(s),
-                                    ))),
-                                ..
-                            },
-                    }) if s == i => continue,
-                    _ => return false,
-                },
-                MacroPattern::FloatLiteral(f) => match val {
-                    ExprKind::Atom(Atom {
-                        syn:
-                            SyntaxObject {
-                                ty: TokenType::Number(NumberLiteral::Real(RealLiteral::Float(s))),
-                                ..
-                            },
-                    }) if s == f => continue,
-                    _ => return false,
-                },
-                MacroPattern::CharacterLiteral(c) => match val {
-                    ExprKind::Atom(Atom {
-                        syn:
-                            SyntaxObject {
-                                ty: TokenType::CharacterLiteral(s),
-                                ..
-                            },
-                    }) if s == c => continue,
-                    _ => return false,
-                },
-                MacroPattern::StringLiteral(s) => match val {
-                    ExprKind::Atom(Atom {
-                        syn:
-                            SyntaxObject {
-                                ty: TokenType::StringLiteral(b),
-                                ..
-                            },
-                    }) if s == b => continue,
-                    _ => return false,
-                },
-                MacroPattern::QuotedExpr(q) => {
-                    // println!("MATCHING QUOTED EXPR: {}", q);
-                    match val {
-                        ExprKind::Quote(boxed_q) if q == boxed_q => continue,
-                        _ => {
-                            return false;
-                        }
-                    }
+pub fn match_vec_pattern(
+    mut patterns: &[MacroPattern],
+    mut list: &[ExprKind],
+    top_level: bool,
+) -> bool {
+    if top_level {
+        list = &list[1..];
+        patterns = &patterns[1..];
+    }
+
+    let mut exprs_iter = list.iter();
+
+    let len_matches = patterns.len() == list.len() || {
+        let has_ellipsis = patterns.iter().any(|pat| pat.is_many());
+
+        has_ellipsis && list.len() >= (patterns.len() - 1)
+    };
```

Comment:
:warning: I just realized that this check is duplicated/redundant with `match_case()`. I have a subsequent PR ready right after this one so, if that's OK, I'd leave this as is for now.